### PR TITLE
chore: upgrade to jest 25

### PR DESCRIPTION
**Pull request by SimenB**
## Summary

The changes to the test code relate to changes in JSDOM that come with Jest 25:

* Several JSDOM workarounds are no longer needed.
* Several tests made assertions to match incorrect JSDOM behavior (e.g. `setAttribute` calls) that JSDOM has now patched to match browsers.
  * https://codesandbox.io/s/resets-value-of-datetime-input-to-fix-bugs-in-ios-safari-1ppwh
* JSDOM no longer triggers default actions when dispatching click events. 
  * https://codesandbox.io/s/beautiful-cdn-ugn8f
* JSDOM fixed (jsdom/jsdom#2700) a bug so that calling `focus()` on an already focused element does not dispatch a `FocusEvent`.
* JSDOM now supports passive events.
* JSDOM has improved support for custom CSS properties.
  * But requires jsdom/cssstyle#112 to land to support webkit prefixed properties.

Closes #17012

## Test Plan

`yarn test`

### CODE REVIEW

**Comment by SimenB** [comment]
I can update babel separately and fix the transform errors

**Comment by eps1lon** [comment]
Bad news: custom css properties are still not working using jsdom
good news: This will be resolved in `cssstyle` (https://github.com/jsdom/cssstyle/issues/89) and will likely not need jsdom 16.

**Comment by eps1lon** [comment]
@SimenB https://github.com/SimenB/react/pull/837 should include all fixes to make JSDOM happy. 

@trueadm might want to take a look at the change Drag-Test.internal

**Comment by trueadm** [comment]
Can you link me to the change you’re referring to? Thanks :)

**Comment by SimenB** [comment]
@eps1lon @trueadm I merged in the changes into this PR

File: packages/react-interactions/events/src/dom/__tests__/Drag-test.internal.js [hunk 1]
```diff
@@ -81,10 +81,8 @@ describe('Drag event responder', () => {
     mouseUpEvent.initEvent('mouseup', true, true);
     divRef.current.dispatchEvent(mouseUpEvent);
 
-    expect(events).toHaveLength(2);
-    expect(events).toEqual(
-      expect.arrayContaining([expect.objectContaining({isChanged: true})]),
-    );
+    expect(events).toHaveLength(0);
+    expect(events).toEqual([]);
```
Comment (trueadm) [inline packages/react-interactions/events/src/dom/__tests__/Drag-test.internal.js]: This isn't right. This test shouldn't have changed.

File: packages/react-interactions/events/src/dom/__tests__/Drag-test.internal.js [hunk 2]
```diff
@@ -138,7 +136,7 @@ describe('Drag event responder', () => {
     mouseUpEvent.initEvent('mouseup', true, true);
     divRef.current.dispatchEvent(mouseUpEvent);
 
-    expect(events).toEqual(['dragstart', 'dragend']);
+    expect(events).toEqual(['dragstart']);
```
Comment (trueadm) [inline packages/react-interactions/events/src/dom/__tests__/Drag-test.internal.js]: Nor should this test.

File: packages/react-interactions/events/src/dom/__tests__/Drag-test.internal.js [hunk 3]
```diff
@@ -188,18 +186,7 @@ describe('Drag event responder', () => {
...
-          diffY: 2,
-        }),
-        expect.objectContaining({
-          diffX: 21,
-          diffY: 21,
-        }),
-      ]),
-    );
+    expect(events).toHaveLength(0);
+    expect(events).toEqual([]);
```
Comment (trueadm) [inline packages/react-interactions/events/src/dom/__tests__/Drag-test.internal.js]: Nor should this test.

**Comment by eps1lon** [comment]
> It suggests something is up with JSDOM if they do?

As I said it's probably something with JSDOM supporting passive event listeners (see https://github.com/SimenB/react/pull/837#discussion_r370646723)

**Comment by trueadm** [comment]
Let me look at changing the Drag tests quickly. Maybe that will help.

File: packages/react-dom/src/events/__tests__/ChangeEventPlugin-test.internal.js [hunk 4]
```diff
@@ -250,9 +250,7 @@ describe('ChangeEventPlugin', () => {
     input.checked = true;
     // Under the hood, uncheck the box so that the click will "check" it again.
     setUntrackedChecked.call(input, false);
-    input.dispatchEvent(
-      new MouseEvent('click', {bubbles: true, cancelable: true}),
-    );
+    input.click();
```
Comment (necolas) [inline packages/react-dom/src/events/__tests__/ChangeEventPlugin-test.internal.js]: Why did these events need to be changed?

File: packages/react-dom/src/__tests__/ReactDOMInput-test.js [hunk 5]
```diff
@@ -1579,25 +1579,13 @@ describe('ReactDOMInput', () => {
...
-        'set attribute type',
-        'set attribute min',
-        'set attribute max',
-        'set attribute step',
-        'set property value',
-        'set attribute value',
-        'set attribute checked',
-      ]);
-    }
+    expect(log).toEqual([
```
Comment (necolas) [inline packages/react-dom/src/__tests__/ReactDOMInput-test.js]: Why was this changed?

File: packages/react-dom/src/events/__tests__/SelectEventPlugin-test.js [hunk 6]
```diff
@@ -43,7 +43,6 @@ describe('SelectEventPlugin', () => {
       <input type="text" onMouseDown={function() {}} />,
       container,
     );
-    node.focus();
```
Comment (necolas) [inline packages/react-dom/src/events/__tests__/SelectEventPlugin-test.js]: Why was this removed?

File: packages/react-dom/src/__tests__/ReactDOMInput-test.js (hunk 5 above)
Comment (eps1lon) [reply packages/react-dom/src/__tests__/ReactDOMInput-test.js]: It was a JSDOM workaround that works different in a browser: https://codesandbox.io/s/resets-value-of-datetime-input-to-fix-bugs-in-ios-safari-1ppwh

Comment (trueadm) [reply packages/react-dom/src/__tests__/ReactDOMInput-test.js]: This needs to go back to what it was before and we need to figure out why there's a difference. We can't just remove it.

File: packages/react-dom/src/events/__tests__/SelectEventPlugin-test.js (hunk 6 above)
Comment (eps1lon) [reply packages/react-dom/src/events/__tests__/SelectEventPlugin-test.js]: Calling `focus()` on an already focused element does not dispatch a FocusEvent. This was a bug in JSDOM that was fixed in https://github.com/jsdom/jsdom/pull/2700

**Comment by trueadm** [comment]
It would also be good to know why all the click events were changed from MouseEvents to .click(). Is this another breaking change in JSDOM?

File: packages/react-interactions/events/src/dom/__tests__/Drag-test.internal.js (hunk 1 above)
Comment (trueadm) [reply packages/react-interactions/events/src/dom/__tests__/Drag-test.internal.js]: However, my PR might fix this https://github.com/facebook/react/pull/17906.

File: packages/react-dom/src/events/__tests__/ChangeEventPlugin-test.internal.js (hunk 4 above)
Comment (eps1lon) [reply packages/react-dom/src/events/__tests__/ChangeEventPlugin-test.internal.js]: JSDOM no longer triggers default actions on dispatching an event it seems. There's no spec around that as far as I know. In any case `.click()` works as well in a browser: https://codesandbox.io/s/should-track-radio-button-cousins-in-a-group-thb8k

File: packages/react-dom/src/__tests__/ReactDOMInput-test.js (hunk 5 above)
Comment (eps1lon) [reply packages/react-dom/src/__tests__/ReactDOMInput-test.js]: But it's not actual browser behavior. See linked codesandbox.

**Comment by eps1lon** [comment]
> It would also be good to know why all the click events were changed from MouseEvents to .click(). Is this another breaking change in JSDOM?

I took all the time to no only fix it, but prepare a PR, annoate changes and prepare codesandboxes with actual browser behavior. Could you respond to these statements instead? It's not clear if I didn't communicate them well enough or if you just ignored them.

**Comment by necolas** [comment]
Yeah it's helpful if the context goes into the PR summary. We don't usually expect context to be found in a PR against another repo.

**Comment by trueadm** [comment]
@eps1lon Sorry if I wasn't clear. I wasn't criticising your work – this is all awesome! I believe I'm also missing the context to the original code somewhere – I can't see any codesandboxes or comments in this PR? I don't see anything in this PR's summary about any of this.

I also wrote the above comment, before your above comments to @necolas, which help expand and clarify on points.

**Comment by necolas** [comment]
I updated the PR summary with the relevant context behind the changes

File: packages/react-dom/src/__tests__/ReactDOMServerIntegrationAttributes-test.js [hunk 7]
```diff
@@ -432,36 +430,40 @@ describe('ReactDOMServerIntegration', () => {
           <div
             style={{
               lineClamp: 10,
-              WebkitLineClamp: 10,
```
Comment (necolas) [inline packages/react-dom/src/__tests__/ReactDOMServerIntegrationAttributes-test.js]: Not sure why we have this. `lineClamp` isn't a valid CSS property...

**Comment by necolas** [comment]
If you could please rebase this on top of latest master, it looks good to me.

**Comment by SimenB** [comment]
I won't be at a computer for a few hours, so if you could rebase that'd be sweet. If not I'll get to it later 🙂

File: packages/react-interactions/events/src/dom/__tests__/Drag-test.internal.js (hunk 3 above)
Comment (SimenB) [reply packages/react-interactions/events/src/dom/__tests__/Drag-test.internal.js]: These 2 assertions are the same aren't they?

Comment (trueadm) [reply packages/react-interactions/events/src/dom/__tests__/Drag-test.internal.js]: These tests are removed now anyway, so it doesn't matter :)

**Comment by SimenB** [comment]
Never got to it yesterday, but rebased now! 👍

**Comment by gaearon** [comment]
Wow this is great, thank you! Leaving for @necolas to merge.

**Comment by acdlite** [comment]
Heads up, I'm reverting this in https://github.com/facebook/react/pull/18376 because of a Jest bug that is fixed in master but unreleased: https://github.com/facebook/jest/pull/9575
